bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+pub mod api;
+mod envelope;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+mod types;
+
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod test;
+
+use codec::{Decode, Encode};
+use envelope::Envelope;
+use frame_support::{
+	traits::{
+		fungible::{Inspect, Mutate},
+		tokens::Balance,
+	},
+	weights::WeightToFee,
+	PalletError,
+};
+use frame_system::{ensure_signed, pallet_prelude::*};
+use scale_info::TypeInfo;
+use snowbridge_core::{
+	inbound::{Message, VerificationError, Verifier},
+	sparse_bitmap::SparseBitmap,
+	BasicOperatingMode,
+};
+use snowbridge_router_primitives::inbound::v2::{
+	ConvertMessage, ConvertMessageError, Message as MessageV2,
+};
+use sp_core::H160;
+use sp_std::vec;
+use types::Nonce;
+pub use weights::WeightInfo;
+use xcm::prelude::{send_xcm, Junction::*, Location, SendError as XcmpSendError, SendXcm, *};
+
+#[cfg(feature = "runtime-benchmarks")]
+use snowbridge_beacon_primitives::BeaconHeader;
+
+pub use pallet::*;
+
+pub const LOG_TARGET: &str = "snowbridge-inbound-queue:v2";
+
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+type BalanceOf<T> =
+	<<T as pallet::Config>::Token as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use frame_support::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[cfg(feature = "runtime-benchmarks")]
+	pub trait BenchmarkHelper<T> {
+		fn initialize_storage(beacon_header: BeaconHeader, block_roots_root: H256);
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The verifier for inbound messages from Ethereum.
+		type Verifier: Verifier;
+		/// XCM message sender.
+		type XcmSender: SendXcm;
+		/// Address of the Gateway contract.
+		#[pallet::constant]
+		type GatewayAddress: Get<H160>;
+		type WeightInfo: WeightInfo;
+		/// Convert a weight value into deductible balance type.
+		type WeightToFee: WeightToFee<Balance = BalanceOf<Self>>;
+		/// AssetHub parachain ID.
+		type AssetHubParaId: Get<u32>;
+		/// Convert a command from Ethereum to an XCM message.
+		type MessageConverter: ConvertMessage;
+		/// Used to burn fees from the origin account (the relayer), which will be teleported to AH.
+		type Token: Mutate<Self::AccountId> + Inspect<Self::AccountId>;
+		/// Used for the dry run API implementation.
+		type Balance: Balance + From<u128>;
+		#[cfg(feature = "runtime-benchmarks")]
+		type Helper: BenchmarkHelper<Self>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A message was received from Ethereum
+		MessageReceived {
+			/// The message nonce
+			nonce: u64,
+			/// ID of the XCM message which was forwarded to the final destination parachain
+			message_id: [u8; 32],
```

Comment:
Does this match the XCM's inner "topic"? If not, let's also add `xcm_topic` to the event for easier flow tracking